File: src/ui/src/common/js/Validator.js
/*
 * Tencent is pleased to support the open source community by making 蓝鲸 available.
 * Copyright (C) 2017-2018 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations under the License.
 */

// import VeeValidate, {Validator} from 'vee-validate'
import VeeValidate, {Validator} from 'vee-validate'
/*
    名称
*/
const isName = {
    getMessage (field, args) { // 錯誤提示
        return '请输入合法的' + field
    },
    validate: value => { // 驗證規則
        return /^([a-zA-Z0-9_ ]|[\u4e00-\u9fa5]|[\uac00-\ud7ff]|[\u0800-\u4e00]){1,20}$/.test(value)
    }
}
Validator.extend('name', isName)
/*
    id
*/
const isID = {
    getMessage (field, args) { // 错误提示
        // zh_CN: field => '请输入' + field
        return '请输入' + field
    },
    validate: value => { // 验证规则
        var reg = /^[a-z0-9_]{1,20}$/
        return reg.test(value)
    }
}
Validator.extend('id', isID)

const isHttp = {
    getMessage (field, args) { // 错误提示
        // zh_CN: field => '请输入' + field
        return '请输入' + field
    },
    validate: value => { // 验证规则
        var reg = /^http:\/\/[^\s]+/
        return reg.test(value)
    }
}
Validator.extend('http', isHttp)
/*
    长字符
*/
const longchar = {
    getMessage (field, args) { // 错误提示
        return '请输入' + field
    },
    validate: value => { // 验证规则
        return /^([a-zA-Z0-9]|[\u4e00-\u9fa5]|[\(\)\+\-《》_,，；;“”‘’。\."'\\\/: ]){0,2000}$/.test(value)
    }
}
Validator.extend('longchar', longchar)

/*
    短字符
*/
const singlechar = {
    getMessage (field, args) { // 错误提示
        return '请输入' + field
    },
    validate: value => { // 验证规则
        return /^([a-zA-Z0-9]|[\u4e00-\u9fa5]|[\(\)\+\-《》_,，；;“”‘’。\."'\\\/: ]){0,256}$/.test(value)
    }
}
Validator.extend('singlechar', singlechar)

/*
    更新Validator
*/
const dictionary = {
    zh_CN: {
        messages: {
            name: () => '请输入正确的内容',
            longchar: () => '请输入正确的内容',
            singlechar: () => '请输入正确的内容',
            id: () => '格式不正确，只能包含下划线，数字，英文小写',
            http: () => '请输入以http://开头的URL',
            required: (field) => '请输入' + field,
            numeric: (field) => '请输入数字',
            regex: (field) => field + '不合法'
        },
        attributes: {
            // name: '中文名',
            // name: '英文名'
            // lifeCycle: '生命周期',
            // maintainers: '运维人员',
            // strictCharacter: '严格字符',
            // looseCharacter: '宽松字符',
            // number: '数字',
            // ipAddress: 'ip地址',
            // commonName: '名称'
        }
    },
    en: {
        messages: {
            name: () => 'Please enter the correct content',
            longchar: () => 'Please enter the correct content',
            singlechar: () => 'Please enter the correct content',
            id: () => 'The format is incorrect and can only contain underscores, numbers, and lowercase English',
            http: () => 'Please enter a URL beginning with http://',
            // lifeCycle: () => '请输入格式为 yyyy-MM-dd 的生命周期',
            // maintainers: () => '请选择运维人员',
            // strictCharacter: () => '请输入正确的验证字符',
            // looseCharacter: () => '请输入正确的宽松字符',
            // number: () => '请输入正确的数字',
            // ipAddress: () => '请输入正确的IP地址',
            // commonName: () => '请输入正确的名称',
            required: (field) => 'Please enter ' + field,
            numeric: (field) => 'Please enter the number',
            regex: (field) => field + ' is illegal'
        },
        attributes: {
            Description: 'Description',
            Name: 'Name'
            // lifeCycle: '生命周期',
            // maintainers: '运维人员',
            // strictCharacter: '严格字符',
            // looseCharacter: '宽松字符',
            // number: '数字',
            // ipAddress: 'ip地址',
            // commonName: '名称'
        }
    }
}
Validator.localize(dictionary)

export default dictionary

File: src/source_controller/coreservice/core/instances/validator.go
/*
 * Tencent is pleased to support the open source community by making 蓝鲸 available.,
 * Copyright (C) 2017-2018 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the MIT License (the ",License",); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 * http://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an ",AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */

package instances

import (
	"configcenter/src/common"
	"configcenter/src/common/blog"
	"configcenter/src/common/errors"
	"configcenter/src/common/http/rest"
	"configcenter/src/common/language"
	"configcenter/src/common/metadata"
)

type validator struct {
	errIf         errors.DefaultCCErrorIf
	properties    map[string]metadata.Attribute
	idToProperty  map[int64]metadata.Attribute
	propertySlice []metadata.Attribute
	require       map[string]bool
	requireFields []string
	uniqueAttrs   []metadata.ObjectUnique
	dependent     OperationDependences
	objID         string
	language      language.CCLanguageIf
}

// NewValidator TODO
// Init init
func NewValidator(kit *rest.Kit, dependent OperationDependences, objID string, bizID int64, language language.CCLanguageIf) (*validator, error) {
	valid := &validator{}
	valid.properties = make(map[string]metadata.Attribute)
	valid.idToProperty = make(map[int64]metadata.Attribute)
	valid.propertySlice = make([]metadata.Attribute, 0)
	valid.require = make(map[string]bool)
	valid.requireFields = make([]string, 0)
	valid.uniqueAttrs = make([]metadata.ObjectUnique, 0)
	valid.objID = objID
	valid.errIf = kit.CCError
	valid.dependent = dependent
	valid.language = language

	result, err := dependent.SelectObjectAttWithParams(kit, objID, []int64{bizID})
	if nil != err {
		return nil, err
	}
	for _, attr := range result {
		valid.properties[attr.PropertyID] = attr
		valid.idToProperty[attr.ID] = attr
		valid.propertySlice = append(valid.propertySlice, attr)
		if attr.IsRequired {
			valid.require[attr.PropertyID] = true
			valid.requireFields = append(valid.requireFields, attr.PropertyID)
		}
	}

	uniqueAttrs, err := valid.dependent.SearchUnique(kit, valid.objID)
	if nil != err {
		return nil, err
	}

	// process model do not have the unique rules, so we ignore it's attribute's unique check
	if len(uniqueAttrs) == 0 && objID != common.BKProcessObjectName {
		blog.Errorf("[validUpdateUnique] there're not unique constraint for %s, return, rid: %s", valid.objID, kit.Rid)
		uniqueAttrs = make([]metadata.ObjectUnique, 0)
	}
	valid.uniqueAttrs = uniqueAttrs

	return valid, nil
}

// NewValidators init validators, returns mapping of biz id to corresponding validator
func NewValidators(kit *rest.Kit, dependent OperationDependences, objID string, bizIDs []int64,
	language language.CCLanguageIf) (map[int64]*validator, error) {

	uniqueAttrs, err := dependent.SearchUnique(kit, objID)
	if err != nil {
		return nil, err
	}

	// process model do not have the unique rules, so we ignore it's attribute's unique check
	if len(uniqueAttrs) == 0 && objID != common.BKProcessObjectName {
		blog.Errorf("there's no unique constraint for %s, rid: %s", objID, kit.Rid)
		uniqueAttrs = make([]metadata.ObjectUnique, 0)
	}

	attributes, err := dependent.SelectObjectAttributes(kit, objID, bizIDs)
	if err != nil {
		return nil, err
	}

	attrMap := make(map[int64][]metadata.Attribute)
	for _, attr := range attributes {
		attrMap[attr.BizID] = append(attrMap[attr.BizID], attr)
	}

	bizValidatorMap := make(map[int64]*validator)
	for _, bizID := range bizIDs {
		if bizValidatorMap[bizID] != nil {
			continue
		}

		validator := &validator{
			properties:    make(map[string]metadata.Attribute),
			idToProperty:  make(map[int64]metadata.Attribute),
			propertySlice: make([]metadata.Attribute, 0),
			require:       make(map[string]bool),
			requireFields: make([]string, 0),
			uniqueAttrs:   uniqueAttrs,
			objID:         objID,
			errIf:         kit.CCError,
			dependent:     dependent,
			language:      language,
		}

		// the instances in biz has both biz attributes and global attributes that has no biz id
		attrArr := attrMap[0]
		if bizID != 0 {
			attrArr = append(attrArr, attrMap[bizID]...)
		}
		for _, attr := range attrArr {
			validator.properties[attr.PropertyID] = attr
			validator.idToProperty[attr.ID] = attr
			validator.propertySlice = append(validator.propertySlice, attr)
			if attr.IsRequired {
				validator.require[attr.PropertyID] = true
				validator.requireFields = append(validator.requireFields, attr.PropertyID)
			}
		}

		bizValidatorMap[bizID] = validator
	}

	return bizValidatorMap, nil
}
